utils/wasm-gen/src/generator/syscalls/invocator.rs

```diff
@@ -310,57 +377,152 @@ impl<'a, 'b> SysCallsInvocator<'a, 'b> {
+                        self.unstructured.int_in_range(0..=upper_limit)?
+                    };
 
-                    let pointer_beyond = self.unstructured.int_in_range(0..=upper_limit)?;
-                    let offset = self.unstructured.int_in_range(0..=pointer_beyond)?;
+                    let length = if let Some(length_allowed_values) = length_allowed_values {
+                        length_allowed_values.get_i32(self.unstructured)?
+                    } else {
+                        self.unstructured.int_in_range(0..=(upper_limit - offset))?
+                    };
 
-                    let first = Instruction::I32Const(offset as i32);
-                    let second = Instruction::I32Const((pointer_beyond - offset) as i32);
-                    log::trace!("  ----  Memory array {first}, {second}");
+                    log::trace!("  ----  Memory array {offset}, {length}");
 
-                    instructions.push(first);
-                    instructions.push(second);
+                    setters.push(ParamSetter::new_i32(offset));
+                    setters.push(ParamSetter::new_i32(length));
                 }
-                ProcessedSysCallParams::MemoryPtrValue => {
-                    // Subtract a bit more so entities from `gsys` fit.
-                    let upper_limit = mem_size.saturating_sub(100);
-                    let offset = self.unstructured.int_in_range(0..=upper_limit)?;
+                ProcessedSysCallParams::MemoryPtrValue { allowed_values } => {
+                    let offset = if let Some(allowed_values) = allowed_values {
+                        allowed_values.get_i32(self.unstructured)?
+                    } else {
+                        // Subtract a bit more so entities from `gsys` fit.
+                        let upper_limit = mem_size.saturating_sub(100);
+                        self.unstructured.int_in_range(0..=upper_limit)? as i32
+                    };
 
-                    let instr = Instruction::I32Const(offset as i32);
-                    log::trace!("  ----  Memory pointer value - {instr}");
+                    let setter = ParamSetter::new_i32(offset);
+                    log::trace!("  ----  Memory pointer value - {offset}");
 
-                    instructions.push(instr);
+                    setters.push(setter);
                 }
             }
         }
 
-        instructions.push(Instruction::Call(call_indexes_handle as u32));
-        instructions.extend(results);
-
         log::trace!(
-            "  ----  Random data after SysCallsInvocator::build_call - {}",
+            "  ----  Random data after SysCallsInvocator::build_param_setters - {}",
             self.unstructured.len()
         );
 
-        Ok(instructions)
+        assert_eq!(setters.len(), params.len());
+
+        Ok(setters)
+    }
+
+    fn build_result_processing_ignored(signature: SysCallSignature) -> Vec<Instruction> {
+        std::iter::repeat(Instruction::Drop)
```

Comment:
We already has `iter` import in `use std::{collections::BTreeMap, iter}`
```suggestion
        iter::repeat(Instruction::Drop)
```